src/noit_metric_tag_search.c

```diff
@@ -65,21 +75,58 @@ static tls_state_t *tls_state_alloc(void);
 static pcre_jit_stack *tls_state_get_pcre_jit_stack(void *arg);
 static pthread_key_t tls_state_key;
 
-static re_matcher_t *
-re_matcher_alloc(char *pattern)
+static bool re_matcher_compile(re_matcher_t *rem)
 {
-  const char *error;
-  re_matcher_t *rem = malloc(sizeof(*rem));
-
-  rem->re_str = pattern;
-  rem->re = pcre_compile(pattern, 0, &error, &rem->error_offset, NULL);
-
+  pcre *re = NULL;
+  if((re = ck_pr_load_ptr(&rem->re)) == NULL && rem->error_offset < 0) {
```

Comment:
no need for atomics, the cache is thread local

Reply:
The cache is thread local, the item in the cache is not necessarily single thread.  It can (and is) passed to other threads.  Now that the re is compiled on demand, we need to use atomics to not race/leak on the compile and assignment.